### ISSUE TITLE
Extmarks are not restored on redo

### ISSUE THREAD

**Issue by ddickstein** [neovim/neovim#30331]
### Problem

Extmarks are not restored properly on redo.

### Steps to reproduce

Run this lua script to set up the buffer.
```lua
vim.cmd "enew"
vim.api.nvim_buf_set_lines(0, 0, -1, 1, {"foobar"})
local ns = vim.api.nvim_create_namespace("")
vim.api.nvim_buf_set_extmark(0, ns, 0, 3, { end_col = 6, hl_group = "Error" })
```
1. Observe restoration works properly on undo
    1. Source the provided lua script
    2. Press `D` (delete "foobar")
    3. Press `u`(undo deletion)
    4. Observe that the highlighting is still present
2. Observe restoration does not work on redo
    1. Source the provided lua script
    2. Press `u` (undo "foobar")
    3. Press `<C-r>` (redo "foobar")
    4. Observe that the highlighting is absent

### Neovim version (nvim -v)

0.10.1

**Comment by clason** [comment]
Possible duplicate/related: https://github.com/neovim/neovim/issues/30184

**Comment by justinmk** [comment]
I guess `undo_restore` is not mentioned because it defaults to true.

Related:
- https://github.com/neovim/neovim/issues/5754

**Comment by zshuzh** [comment]
I think this is _technically_ a feature request and not a bug. The very last paragraph in `:h extmarks` states:

>  Extmark positions changed by an edit will be restored on undo/redo. Creating and deleting extmarks is not a buffer change, thus new undo states are not created for extmark changes

In your first example, an extmark was created, and then its position was then changed by an edit (delete "foobar"), so as stated above, the extmark is restored on undo.

However, in your second example, an edit was made (add "foobar"), and _then_ an extmark was added, so as stated above, a new undo state will not be created for this extmark change.

An argument could be made that the "undo" in the second example should count as an edit and the extmark position should be restored on redo. I think this is also what would intuitively be expected. A possible implementation would be to modify extmark information associated with the relevant undo state (contained in `u_header.uh_extmark`). This could also be put behind a `redo_restore` option to `nvim_buf_set_extmark` to match the existing `undo_restore` option.

I'm happy to take a stab at implementing this, but I'd appreciate it if a maintainer could weigh in and let me know if there's any reason not to implement this feature or if there's a problem with my proposed approach. @justinmk I hope you don't mind me pinging you, but do you have any thoughts?

**Comment by bfredl** [comment]
Yeah regardless if a bug or a feature request I think the expected behavior makes more sense. PR welcome.

> This could also be put behind a redo_restore option to nvim_buf_set_extmark to match the existing undo_restore option.

No need for a separate flag. `undo_restore` refers to the entire undo machinery, including redoing, branches (as far as possible) etc.
